File: Aula AG 1/cromossomo.py
import random
from util import Util

class Cromossomo:
    def __init__(self, rota):
        self.rota = rota
        self.aptidao = self.calcular_aptidao()

    def calcular_aptidao(self):
        penalidade = 0
        cidades_vistas = {}

        # Penalidade por inversões
        for i in range(len(self.rota)):
            for j in range(i+1, len(self.rota)):
                if self.rota[i] > self.rota[j]:
                    penalidade += 10

        # Penalidade por cidades repetidas
        for cidade in self.rota:
            cidades_vistas[cidade] = cidades_vistas.get(cidade, 0) + 1

        for cidade, count in cidades_vistas.items():
            if count > 1:
                # Cada par repetido adiciona 20
                pares_repetidos = (count * (count - 1)) // 2
                penalidade += 20 * pares_repetidos

        return penalidade

    def __str__(self):
        return f'Rota: {"".join(map(str, self.rota))} | Penalidade: {self.aptidao}'

    @staticmethod
    def gerar_populacao(populacao, tamanho_populacao):
        for _ in range(tamanho_populacao):
            rota = Util.gerar_rota()
            populacao.append(Cromossomo(rota))

    @staticmethod
    def exibir_populacao(populacao):
        for individuo in populacao:
            print(individuo)

    @staticmethod
    def selecionar_por_torneio(populacao, nova_populacao, taxa_selecao):
        torneio = []
        qtd_selecionados = int(taxa_selecao * len(populacao) / 100)
        nova_populacao.append(populacao[0])  # elitismo

        i = 1
        while i <= qtd_selecionados:
            c1 = random.choice(populacao)
            c2 = random.choice([c for c in populacao if c != c1])
            c3 = random.choice([c for c in populacao if c != c1 and c != c2])

            torneio = [c1, c2, c3]
            torneio.sort(key=lambda cromossomo: cromossomo.aptidao)  # Menor penalidade = melhor

            selecionado = torneio[0]
            if selecionado not in nova_populacao:
                nova_populacao.append(selecionado)
                i += 1

    @staticmethod
    def reproduzir(populacao, nova_populacao, taxa_reproducao):
        qtd_reproduzidos = int(taxa_reproducao * len(populacao) / 100)
        i = 0
        while i < qtd_reproduzidos:
            pai = random.choice(populacao)
            mae = random.choice([c for c in populacao if c != pai])

            ponto_corte = len(pai.rota) // 2
            filho1_rota = pai.rota[:ponto_corte] + mae.rota[ponto_corte:]
            filho2_rota = mae.rota[:ponto_corte] + pai.rota[ponto_corte:]

            nova_populacao.append(Cromossomo(filho1_rota))
            nova_populacao.append(Cromossomo(filho2_rota))
            i += 2

        # podar caso tenha ultrapassado
        while len(nova_populacao) > len(populacao):
            nova_populacao.pop()

    @staticmethod
    def mutar(populacao):
        qtd_mutantes = random.randint(1, int(len(populacao) / 5))
        for _ in range(qtd_mutantes):
            pos = random.randrange(len(populacao))
            mutante = populacao[pos]

            rota_mutada = mutante.rota[:]
            idx1 = random.randint(0, len(rota_mutada) - 1)
            idx2 = random.randint(0, len(rota_mutada) - 1)
            rota_mutada[idx1], rota_mutada[idx2] = rota_mutada[idx2], rota_mutada[idx1]

            populacao[pos] = Cromossomo(rota_mutada)

File: Atividade cidades java/ag/Cromossomo.java
package ag;

import java.util.ArrayList;
//import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class Cromossomo {
    private final int[] rota;
    private final int aptidao;

    public Cromossomo(int[] rota) {
        this.rota = rota;
        this.aptidao = calcularAptidao();
    }

    private int calcularAptidao() {
        int penalidade = 0;
        
        // Penalidade por inversões
        for (int i = 0; i < rota.length; i++) {
            for (int j = i + 1; j < rota.length; j++) {
                if (rota[i] > rota[j]) {
                    penalidade += 10;
                }
            }
        }

        // Penalidade por cidades repetidas
        Map<Integer, Integer> cidadesVistas = new HashMap<>();
        for (int cidade : rota) {
            cidadesVistas.put(cidade, cidadesVistas.getOrDefault(cidade, 0) + 1);
        }

        for (int count : cidadesVistas.values()) {
            if (count > 1) {
                int paresRepetidos = (count * (count - 1)) / 2;
                penalidade += 20 * paresRepetidos;
            }
        }

        return penalidade;
    }

    public int[] getRota() {
        return rota;
    }

    public int getAptidao() {
        return aptidao;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int cidade : rota) {
            sb.append(cidade);
        }
        return "Rota: " + sb.toString() + " | Penalidade: " + aptidao;
    }

    public static void gerarPopulacao(List<Cromossomo> populacao, int tamanho) {
        for (int i = 0; i < tamanho; i++) {
            populacao.add(new Cromossomo(Util.gerarRota()));
        }
    }

    public static void exibirPopulacao(List<Cromossomo> populacao) {
        for (Cromossomo c : populacao) {
            System.out.println(c);
        }
    }

    public static void selecionarPorTorneio(List<Cromossomo> populacao, List<Cromossomo> novaPopulacao, int taxaSelecao) {
        Random random = new Random();
        int qtdSelecionados = (taxaSelecao * populacao.size()) / 100;
        novaPopulacao.add(populacao.get(0)); // Elitismo

        int i = 1;
        while (i <= qtdSelecionados) {
            Cromossomo c1 = populacao.get(random.nextInt(populacao.size()));
            Cromossomo c2 = populacao.get(random.nextInt(populacao.size()));
            while (c2 == c1) {
                c2 = populacao.get(random.nextInt(populacao.size()));
            }
            Cromossomo c3 = populacao.get(random.nextInt(populacao.size()));
            while (c3 == c1 || c3 == c2) {
                c3 = populacao.get(random.nextInt(populacao.size()));
            }

            List<Cromossomo> torneio = new ArrayList<>();
            torneio.add(c1);
            torneio.add(c2);
            torneio.add(c3);
            torneio.sort(Comparator.comparingInt(Cromossomo::getAptidao));

            Cromossomo selecionado = torneio.get(0);
            if (!novaPopulacao.contains(selecionado)) {
                novaPopulacao.add(selecionado);
                i++;
            }
        }
    }

    public static void reproduzir(List<Cromossomo> populacao, List<Cromossomo> novaPopulacao, int taxaReproducao) {
        Random random = new Random();
        int qtdReproduzidos = (taxaReproducao * populacao.size()) / 100;
        int i = 0;

        while (i < qtdReproduzidos) {
            Cromossomo pai = populacao.get(random.nextInt(populacao.size()));
            Cromossomo mae = populacao.get(random.nextInt(populacao.size()));
            while (mae == pai) {
                mae = populacao.get(random.nextInt(populacao.size()));
            }

            int pontoCorte = pai.getRota().length / 2;
            
            int[] filho1 = new int[pai.getRota().length];
            System.arraycopy(pai.getRota(), 0, filho1, 0, pontoCorte);
            System.arraycopy(mae.getRota(), pontoCorte, filho1, pontoCorte, mae.getRota().length - pontoCorte);

            int[] filho2 = new int[mae.getRota().length];
            System.arraycopy(mae.getRota(), 0, filho2, 0, pontoCorte);
            System.arraycopy(pai.getRota(), pontoCorte, filho2, pontoCorte, pai.getRota().length - pontoCorte);

            novaPopulacao.add(new Cromossomo(filho1));
            novaPopulacao.add(new Cromossomo(filho2));
            i += 2;
        }

        while (novaPopulacao.size() > populacao.size()) {
            novaPopulacao.remove(novaPopulacao.size() - 1);
        }
    }

    public static void mutar(List<Cromossomo> populacao) {
        Random random = new Random();
        int qtdMutantes = random.nextInt(populacao.size() / 5) + 1;

        for (int i = 0; i < qtdMutantes; i++) {
            int pos = random.nextInt(populacao.size());
            Cromossomo mutante = populacao.get(pos);
            int[] novaRota = mutante.getRota().clone();
            
            int idx1 = random.nextInt(novaRota.length);
            int idx2 = random.nextInt(novaRota.length);
            while (idx2 == idx1) {
                idx2 = random.nextInt(novaRota.length);
            }
            
            int temp = novaRota[idx1];
            novaRota[idx1] = novaRota[idx2];
            novaRota[idx2] = temp;
            
            populacao.set(pos, new Cromossomo(novaRota));
        }
    }
}